emergence_lib/src/player_interaction/tile_selection.rs

```diff
@@ -0,0 +1,238 @@
+    pub fn deselect_tile(&mut self, tile_pos: TilePos) {
+        self.cache_selection();
+        self.selection.remove(&tile_pos);
+    }
+
+    /// Selects or unselects a single tile.
+    ///
+    /// If it is not selected, select it.
+    /// If it is already selected, remove it from the selection.
+    pub fn toggle_tile(&mut self, tile_pos: TilePos) {
+        self.cache_selection();
+        if self.selection.contains(&tile_pos) {
+            self.selection.clear();
+        } else {
+            // Clear and then reinsert rather than making a new struct
+            // to avoid a pointless allocation
+            self.selection.clear();
+            self.selection.insert(tile_pos);
+        }
+    }
+
+    /// Toggles the selection of a group of tiles.
+    ///
+    /// For each tile:
+    /// - if it are not selected, select it.
+    /// - if it is already selected, remove it from the selection.
+    pub fn toggle_tiles(&mut self, tile_pos_collection: impl IntoIterator<Item = TilePos>) {
+        self.cache_selection();
+
+        tile_pos_collection.into_iter().for_each(|tile_pos| {
+            if self.selection.contains(&tile_pos) {
+                self.selection.remove(&tile_pos);
+            } else {
+                self.selection.insert(tile_pos);
+            }
+        });
+    }
+
+    /// Adds or removes a tile to the selection.
+    ///
+    /// If it is not selected, select it.
+    /// If it is already selected, remove it from the selection.
+    pub fn modify_selection(&mut self, tile_pos: TilePos) {
+        self.cache_selection();
+        if self.selection.contains(&tile_pos) {
+            self.selection.remove(&tile_pos);
+        } else {
+            self.selection.insert(tile_pos);
+        }
+    }
+
+    /// The current set of selected tiles
+    pub fn selection(&self) -> &HashSet<TilePos> {
+        &self.selection
+    }
+
+    /// The previous set of selected tiles
+    pub fn previous_selection(&self) -> &HashSet<TilePos> {
+        &self.previous_selection
+    }
+
+    /// Stores the current tile selection to be used to compute the set of changed tiles efficiently
+    fn cache_selection(&mut self) {
+        self.previous_selection = self.selection.clone();
+    }
+
+    /// Clears the set of selected tiles.
+    pub fn clear_selection(&mut self) {
+        self.selection.clear();
+    }
+
+    /// The number of selected tiles.
+    pub fn len(&self) -> usize {
+        self.selection.len()
+    }
+
+    /// Are any tiles selected?
+    pub fn is_empty(&self) -> bool {
+        self.selection.is_empty()
+    }
+
+    /// Is the given tile in the selection?
+    pub fn contains(&self, tile_pos: &TilePos) -> bool {
+        self.selection.contains(&tile_pos)
+    }
+
+    /// Compute the set of newly added tiles
+    pub fn added_tiles(&self) -> HashSet<TilePos> {
+        self.selection
+            .difference(self.previous_selection())
+            .into_iter()
+            .map(|p| *p)
```

Comment:
I was surprised to see that there isn't a built in function to this.
Can `collect` dereference automatically maybe?

Reply:
As far as I can tell, we would need to change the output type of the function to a `HashMap<&TilePos>` to .`collect()` the `&TilePos` we produce with the mapping here, do some lifetime hackery to not need to deference here, or [abuse the Deref trait](https://rust-unofficial.github.io/patterns/anti_patterns/deref.html) by implementing `Deref `for `TilePos`.

Reply:
Apparently `.copied()` exists so there is in fact a built in function that can deref for this. @alice-i-cecile does making this `.copied()` instead make sense to you?

Reply:
This method has to copy / clone: look at the signature. It takes a reference and returns an owned value.

.copied is clearer here :)